Abac

andresgaragiola opened the pull request:
We have been work on this feature with @diegomasini  and @raullaprida  it is a large PR including  support to attributes and Attribute Based Access Control mechanisms.  This PR include the changes of the closed PR #1169. 
## Description

We have added the doc aca.md to describe this feature (this document is included in the doc folder but can be acceded from [this](https://github.com/hyperledger/fabric/wiki/Attribute-Based-Access-Control-%28ABAC%29) wiki page too).
We have added an example called "chaincode_example07" this example describe how ABAC verification could be invoked. The document "ChaincodeAPI.md" was modified to explain it.
## Motivation and Context

This is a required feature as is mentioned in #1139
Fixes #1139
## How Has This Been Tested?
- Current behave tests has been run without errors.
- Current unit tests has been run. The test TestExecuteDeploySysChaincode defined in the file exectransaction_test.go fails but this test fails in the master too (we are working with @muralisrini to understand if this is a problem related with the PR or not)
- We have modified the test TestClientGetAttributesFromTCert defined in crypto_test.go to test the new functionality.
- A new behave test is pending we will work with @jeffgarratt on that.
## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have added a [Signed-off-by](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md#legal-stuff) (This is a PR **for review** that is the reason because I haven't added a "Signed-off", I will add a "Signed-off" once the test problem described above will be fixed).
- [X] We added new documentation
- [X] We added new tests
- [X] I have run [goimports](https://godoc.org/golang.org/x/tools/cmd/goimports), [go vet](https://golang.org/cmd/vet/), and [golint](https://github.com/golang/lint). I have cleaned up all valid errors and warnings in code I have added or modified. These tools may generate false positives. Don't be worried about ignoring some errors or warnings. The goal is clean, consistent, and readable code.

Signed-off-by:

Commit:
Revert "Adding read attributes AI"

This reverts commit 4695f6d9dcf317139a97b221e3e87a1296b0d81f.

diegomasini committed:
Merge remote-tracking branch 'upstream/master'

diegomasini committed:
Merge remote-tracking branch 'upstream/master'

diegomasini committed:
Merge remote-tracking branch 'upstream/master'

diegomasini committed:
Merge with upstream

Commit:
First commit of ABAC branch to share with the other team members.

Commit:
Maerge remote-tracking branch 'andresremote/ABAC' into ABAC

Merge con branch de andres

Commit:
prek0 returned

diegomasini committed:
Merge remote-tracking branch 'upstream/master'

Commit:
Adding the first version of Attribute Certificate Authority (ACA).

Commit:
Adding the first version of Attribute Certificate Authority (ACA).

Commit:
Adding ACA to the membership services.

Commit:
Merge remote-tracking branch 'upstream/master' into ABAC

diegomasini committed:
Merge remote-tracking branch 'andres/abac' into abac

Commit:
Attribute Certificate Authority implemented and integrated with TCA and ECA.

Commit:
Fixing error in unit tests crypto_test.go

diegomasini committed:
Added preK0 for attribute key derivation in TCerts

diegomasini committed:
Merge remote-tracking branch 'andres/ABAC' into abac

Commit:
attributes

Commit:
attributes

diegomasini committed:
Added support to specify user attributes from the CLI. Removed previous attribute specification through configuration file

diegomasini committed:
Fixed crypto tests using new attribute request from user support

Commit:
Merge remote-tracking branch 'upstream/master' into ABAC

Conflicts:
	membersrvc/ca/tca.go

andresgaragiola committed:
Merge remote-tracking branch 'origin/ABAC' into abac

Conflicts:
	core/crypto/client_tca.go
	core/crypto/utils/cert.go
	core/devops.go
	membersrvc/ca/tca.go

andresgaragiola committed:
Adding aca_flow.jpg image for documentations propuses.

andresgaragiola committed:
Adding aca.md (documenation about Attribute Certificate Authority.

andresgaragiola committed:
Added ABAC libraries and mechaninms to send attribute keys during invoke.

andresgaragiola committed:
Added ABACVerfier API to verify attribues from the application layer.

andresgaragiola committed:
Fixed issues with example07 verifing attributes.

andresgaragiola committed:
Reveting core.yaml file.

andresgaragiola committed:
Added example 07 showing the use of ABAC.

diegomasini committed:
Merge with upstream

andresgaragiola committed:
Added encryption of attributes header.

diegomasini committed:
Merge remote-tracking branch 'andres/abac' into abac

diegomasini committed:
Merge remote-tracking branch 'andres/abac' into abac

andresgaragiola committed:
Fixed error during behave tests and other during unit tests.

diegomasini committed:
Merge remote-tracking branch 'upstream/master' into abac

diegomasini committed:
Merge remote-tracking branch 'andres/abac' into abac

andresgaragiola committed:
Adding Padding to the attributes.

diegomasini committed:
Merge remote-tracking branch 'andres/abac' into abac

andresgaragiola committed:
Merge remote-tracking branch 'origin/ABAC' into abac

andresgaragiola committed:
Fixing behave tests

andresgaragiola committed:
Fixed ACA documentation.

diegomasini committed:
Merge remote-tracking branch 'andres/abac' into abac

andresgaragiola committed:
Adding chaincode_example_07.go and removing chaincode_example_07 from the repository

diegomasini committed:
Merge remote-tracking branch 'andres/abac' into abac

andresgaragiola committed:
Adding multithread support on TCert generation.

andresgaragiola committed:
Merge remote-tracking branch 'upstream/master' into abac

diegomasini committed:
Merge remote-tracking branch 'andres/abac' into abac

Commit:
Merge remote-tracking branch 'andresremote/abac' into andresMT

ihoiohiohoihio

andresgaragiola committed:
Fixed issues related with multithreading tcert pool.

andresgaragiola committed:
Disabled security, privacy, abac in core.yaml

diegomasini committed:
Merge remote-tracking branch 'upstream/master' into abac

diegomasini committed:
Merge remote-tracking branch 'andres/abac' into abac

Commit:
Chaincode API updated with attributes

diegomasini committed:
Merge remote-tracking branch 'upstream/master' into abac

andresgaragiola committed:
Fixing code format warning.

diegomasini committed:
Merge remote-tracking branch 'andres/abac' into abac

Commit:
Merge remote-tracking branch 'upstream/master' into andresMT

sdsdsd

sdsd

Commit:
Merge remote-tracking branch 'andresremote/abac' into andresMT

sssdsdd

diegomasini committed:
Fixed ca_test NewCA test.

andresgaragiola committed:
Modified aca.md file.

andresgaragiola committed:
Merge remote-tracking branch 'upstream/master' into abac

Conflicts:
	membersrvc/ca/ca.go

andresgaragiola committed:
Merge remote-tracking branch 'upstream/master' into abac

andresgaragiola committed:
Merge remote-tracking branch 'raul/andresMT' into abac

binhn commented:
@andresgaragiola Some comments:
1. In some files, the copyright header is wrong
2. It doesn't look like you run golint, govet (see [coding guidelines](https://github.com/hyperledger/fabric/blob/master/CONTRIBUTING.md)
3. I would add more detail on the chaincode stub public API comments. For example,

```
//Verifies all the attributes included in attrs.
func (stub *ChaincodeStub) VerifyAttributes(attrs...*ac.Attribute)  (bool, error)
```

Not enough info for developers
4. Don't change `ChaincodeAPI.md` file, which points developers to the `godoc`, so you should document the public functions from the code as commented in 3 above
5. `aca.md` should be in `fabric/docs/tech` directory and work with @joshhus to have an appropriate link from `docs/README.md` file. Also, for public APIs document, you should point to the `godoc` (see `ChaincodeAPI.md` as an example)

@muralisrini please review and look at the unit-test failure

muralisrini commented:
@andresgaragiola  @diegomasini  make unit-test succeeded on latest master for me. See attached file
[out.txt](https://github.com/hyperledger/fabric/files/271859/out.txt)

christo4ferris commented on core/crypto/crypto_test.go:
```diff
@@ -59,6 +61,9 @@ var (
 
 	ksPwd = []byte("This is a very very very long pw")
+	
+	attributes = map[string]string{"company":"IBM", "position":"Software Engineer"}
```
please use a fictional company name such as example.com

andresgaragiola replied:
Thank you @christo4ferris I have modified that file now.

andresgaragiola replied:
Thanks @muralisrini your right unit test are running successfully in lasted master and in the abac branch too.  Probably was a temporary error.

christo4ferris commented:
nearly 2700 LOC and only 39 are changes to an existing test? Seriously, where are the tests that support this PR? Behave tests aren't sufficient. There should be functional tests for all the new function included. Further, we should not be submitting PRs this large for any reason.

christo4ferris commented on core/chaincode/shim/crypto/ac/abac.go:
```diff
@@ -0,0 +1,145 @@
+/*
```
should use https://github.com/hyperledger/fabric/blob/master/docs/dev-setup/headers.txt replacing "[COPYRIGHT OWNER]" with "IBM Corp. 2016"

andresgaragiola replied:
@christo4ferris your right we will work adding more tests to this functionality. About the header I have modified this file in a recent commit.

christo4ferris commented on core/crypto/abac/attributes.go:
```diff
@@ -0,0 +1,254 @@
+/*
```
should use https://github.com/hyperledger/fabric/blob/master/docs/dev-setup/headers.txt replacing "[COPYRIGHT OWNER]" with "IBM Corp. 2016"

andresgaragiola replied:
@christo4ferris idem abac.go

christo4ferris commented on core/crypto/abac/proto/abac.proto:
```diff
@@ -0,0 +1,35 @@
+/*
+Licensed to the Apache Software Foundation (ASF) under one
```
should use https://github.com/hyperledger/fabric/blob/master/docs/dev-setup/headers.txt replacing "[COPYRIGHT OWNER]" with "IBM Corp. 2016"

andresgaragiola replied:
@christo4ferris fixed.

christo4ferris commented on examples/chaincode/go/chaincode_example07/chaincode_example07.go:
```diff
@@ -0,0 +1,123 @@
+/*
+Licensed to the Apache Software Foundation (ASF) under one
```
should use https://github.com/hyperledger/fabric/blob/master/docs/dev-setup/headers.txt replacing "[COPYRIGHT OWNER]" with "IBM Corp. 2016"

andresgaragiola replied:
@christo4ferris fixed.

christo4ferris commented on membersrvc/ca/aca.go:
```diff
@@ -0,0 +1,546 @@
+/*
+Licensed to the Apache Software Foundation (ASF) under one
```
should use https://github.com/hyperledger/fabric/blob/master/docs/dev-setup/headers.txt replacing "[COPYRIGHT OWNER]" with "IBM Corp. 2016"

andresgaragiola replied:
@christo4ferris fixed.

andresgaragiola committed:
Fixed issues identified after PR for review.

andresgaragiola committed:
Merge remote-tracking branch 'upstream/master' into abac

srderson commented:
Please do not name example chaincodes chaincode_examplexx. This is bad practice that we want to stop. Just give it a descriptive name indicating what it actually does..

srderson commented on examples/chaincode/go/chaincode_example07/chaincode_example07.go:
```diff
@@ -0,0 +1,123 @@
+type SimpleChaincode struct {
+}
+
+func (t *SimpleChaincode) Init(stub *shim.ChaincodeStub, function string, args []string) ([]byte, error) {
```
Missing comment. You must run golint on all files

andresgaragiola replied:
@srderson thank you for your review I will work in your observations.

srderson commented on examples/chaincode/go/chaincode_example07/chaincode_example07.go:
```diff
@@ -0,0 +1,123 @@
+
+// Run callback representing the invocation of a chaincode
+// This chaincode will manage two accounts A and B and will transfer X units from A to B upon invoke
+func (t *SimpleChaincode) Run(stub *shim.ChaincodeStub, function string, args []string) ([]byte, error) {
```
Why is there a Run function? This has been deprecated. It looks like it was just copy/pasted as the comment doesn't seem to match the chaincode

srderson commented on examples/chaincode/go/chaincode_example07/chaincode_example07.go:
```diff
@@ -0,0 +1,123 @@
+)
+
+// SimpleChaincode example simple Chaincode implementation
+type SimpleChaincode struct {
```
Please add comments indicating the purpose and function of this chaincode.

andresgaragiola committed:
Fixed header on example_chaincode07.go

andresgaragiola commented:
@binhn thank you for your review.
1. In some files, the copyright header is wrong. => **Done**. I have fixed attributes.go, abac.go and example_chaincode07.go.
2. It doesn't look like you run golint, govet (see coding guidelines.  => **Working**. I found some files as attributes.go, abac.go with this issue. I have run the tools on this files now I will be reviewing for other files.
3. I would add more detail on the chaincode stub public API comments. For example. => **Done**.
4. Don't change ChaincodeAPI.md file, which points developers to the godoc, so you should  document the public functions from the code as commented in 3 above. => **Done**, I have removed our changes from this file and put this in the code as comments.
5. aca.md should be in fabric/docs/tech directory and work with @joshhus to have an appropriate link from docs/README.md file. Also, for public APIs document, you should point to the godoc (see ChaincodeAPI.md as an example) => **Done**. I have moved the file to ../docs/tech I will contact @joshhus to see how link this on README.md

andresgaragiola committed:
Fixed header on aca.go and abac.proto.

andresgaragiola committed:
Fixing some format errors in chaincode_example07 and setting crypto loggin level to info.

andresgaragiola committed:
Setting loggin level of crypto to "info" in the file crypto_test.yaml

muralisrini commented:
@andresgaragiola @diegomasini 

Few suggestions/observations.

A nit but good to cleanup 
- "type ChaincodeHolder interface" appears to be a local interface... perhaps change it to "type chaincodeHolder interface". Makes the intent clear and avoids exports (and take golint off your back a bit :-) )

More importantly, adding fabric imports in the shim creates more dependencies between shim and the fabric (e.g.,  "fabric/core/crypto/abac"  in "fabric/core/chaincode/shim/crypto/ac").  This tight coupling gives raise to some core issues
- makes it hard to keep the dependency tree from growing leading to a bloated chaincode image. This is not a problem currently as we include the fabric in the image but will become one soon
- makes it harder to make support for chaincode languages fully equivalent to golang chaincode

(I know there are other imports like this, but why add to the problem when we should be decoupling ?)

One approach could be to rethink design so things are done in the peer (may not always be possible) or duplicate the code in the shim. That would reduce dependencies between shim and the fabric.

Above may be done in different PRs if we agree they have merit.

andresgaragiola committed:
Fixed issues with crypto_test.go that has occurred after last commits.

andresgaragiola committed:
Merge remote-tracking branch 'upstream/master' into abac

Conflicts:
	core/crypto/client_tca.go

diegomasini commented:
@muralisrini thanks for the suggestions, a couple of answers to them:
-  you are right "type ChaincodeHolder interface" is internal so we changed it to "type chaincodeHolder interface".
- agreed on the fabric imports in the shim, you are right here too, it is better to not contribute to that problem, but we prefer to attack that problem in a separated PR. We would like to redesign that part to avoid duplicating code but if there is no other solution we will do that.

andresgaragiola committed:
Modified interface "ChaincodeHolder" to be private, and changed the name of the example to "authorizable_counter".

andresgaragiola commented:
@srderson I have modified the example according to your observations.

andresgaragiola committed:
Renamed aca.md file to abac.md

andresgaragiola committed:
Merge remote-tracking branch 'upstream/master' into abac

Conflicts:
	membersrvc/ca/tca.go

andresgaragiola committed:
Added reference to abac.md in README.md

diegomasini committed:
Disable security, privacy and abac by default

diegomasini committed:
Merge branch 'abac' of https://github.com/andresgaragiola/fabric into abac

srderson commented on core/crypto/abac/proto/abac.proto:
```diff
@@ -0,0 +1,32 @@
+message ABACMetadata {
+    //Original metadata bytes
+    bytes Metadata = 1;
+   	//Entries for each attributes considered.
```
random indent?

srderson commented on core/crypto/client_impl.go:
```diff
@@ -37,74 +37,91 @@ type clientImpl struct {
 	}
 
+      keys := make([]string, len(attributes))
+    for k := range attributes {
```
seems like goimports (which inclides gofmt) needs to be run on this file. Formatting looks off.

srderson commented on core/crypto/client_tca.go:
```diff
@@ -190,17 +189,17 @@ func (client *clientImpl) getTCertFromExternalDER(der []byte) (tCert, error) {
-			client.enrollPrivKey.PublicKey.X, client.enrollPrivKey.PublicKey.Y,
-			tempX, tempY,
-		)
+			tempSK.PublicKey.Add(
```
looks like goimports (which includes gofmt) needs to be run

andresgaragiola replied:
@srderson I have checked this file and that is the format generated by goimports.  I tried removing the indent a run goimports again an the indent is generated automatically by goimports.

srderson replied:
Oh, I didn't notice the `=` on line 191. In that case the indent makes sense.

srderson commented on core/crypto/client_tca.go:
```diff
@@ -556,19 +565,30 @@ func (client *clientImpl) getTCertsFromTCA(num int) error {
+		tcertAttr.AttributeValue = value
+
+		attributesList = append(attributesList, tcertAttr)
+		//attributesList = append(attributesList, &membersrvc.TCertAttribute{k, attributes[k]})
```
any reason to keep this comment here?

andresgaragiola replied:
Removed

srderson commented on core/crypto/client_tcert.go:
```diff
@@ -18,27 +18,40 @@ package crypto
 	return tCert.cert
 }
 
+func (tCert *tCertImpl) GetPreK0() []byte {
```
Missing comment on public function. golint needs to be run

andresgaragiola replied:
Done

srderson commented on core/crypto/client_tcert.go:
```diff
@@ -18,27 +18,40 @@ package crypto
+}
+
+
 func (tCert *tCertImpl) Sign(msg []byte) ([]byte, error) {
```
Missing comment on public function. golint needs to be run

andresgaragiola replied:
Done

srderson commented on core/crypto/client_tcert.go:
```diff
@@ -57,3 +70,11 @@ func (tCert *tCertImpl) Verify(signature, msg []byte) (err error) {
 	return
 }
+
+func (tCert *tCertImpl) GetKForAttribute(attributeName string) ([]byte, error) {
```
Missing comment on public function. golint needs to be run

andresgaragiola replied:
Added comments

srderson commented on core/crypto/client_tcert_pool_mt.go:
```diff
@@ -18,186 +18,299 @@ package crypto
+	entriesMutex *sync.Mutex
+}
+
+func (tCertPool *tCertPoolMultithreadingImpl) Start() (err error) {
```
Missing comment on public function. golint needs to be run

andresgaragiola replied:
Added comments

srderson commented on core/crypto/client_tcert_pool_mt.go:
```diff
@@ -18,186 +18,299 @@ package crypto
+	runtime.Gosched()
+}
+
+func (tCertPool *tCertPoolMultithreadingImpl) Stop() (err error) {
```
Missing comment on public function. golint needs to be run

andresgaragiola replied:
Added comments

srderson commented on core/crypto/client_tcert_pool_mt.go:
```diff
@@ -18,186 +18,299 @@ package crypto
+	return poolEntry, nil
+}
+
+func (tCertPool *tCertPoolMultithreadingImpl) GetNextTCert(attributes map[string]string) (tCertBlock *TCertBlock, err error) {
```
Missing comment on public function. golint needs to be run

andresgaragiola replied:
Added comment

srderson commented on core/crypto/client_tcert_pool_mt.go:
```diff
@@ -18,186 +18,299 @@ package crypto
+	return poolEntry.GetNextTCert(attributes)
+}
+
+func (tCertPool *tCertPoolMultithreadingImpl) AddTCert(tCertBlock *TCertBlock) (err error) {
```
Missing comment on public function. golint needs to be run

andresgaragiola replied:
Added comment

srderson commented on core/crypto/client_tcert_pool_st.go:
```diff
@@ -19,13 +19,33 @@ package crypto
 )
 
+
+type TCertBlock struct {
```
Missing comment on public struct. golint needs to be run

andresgaragiola replied:
Added comment

srderson commented on core/crypto/client_tcert_pool_st.go:
```diff
@@ -19,13 +19,33 @@ package crypto
+	attributesHash string
+}
+
+type TCertDBBlock struct {
```
Missing comment on public struct. golint needs to be run

andresgaragiola replied:
Added comment

srderson commented on core/crypto/client_tcert_pool_st.go:
```diff
@@ -67,49 +87,119 @@ func (tCertPool *tCertPoolSingleThreadImpl) Stop() (err error) {
 	defer tCertPool.m.Unlock()
 
-	tCertPool.client.debug("Found %d unused TCerts...", tCertPool.len)
+	//tCertPool.client.debug("Found %d unused TCerts...", tCertPool.len)
```
Any reason to keep this comment?

andresgaragiola replied:
Removed commented line.

srderson commented on core/crypto/client_tcert_pool_st.go:
```diff
@@ -67,49 +87,119 @@ func (tCertPool *tCertPoolSingleThreadImpl) Stop() (err error) {
-func (tCertPool *tCertPoolSingleThreadImpl) GetNextTCert() (tCert tCert, err error) {
+
+
+func CalculateAttributesHash(attributes map[string]string) (attrHash string) {
```
Missing comment on public function. golint needs to be run

andresgaragiola replied:
Added comment

srderson commented on core/rest/rest_api.go:
```diff
@@ -1614,7 +1615,7 @@ func (s *ServerOpenchainREST) processChaincodeInvokeOrQuery(method string, spec
 		// Trigger the chaincode query through the devops service
 		//
-
+		
```
Random tab? Indicates goimports was not run

andresgaragiola replied:
I have formatted this code with goimports but indent was not modified.

srderson commented on examples/chaincode/go/authorizable_counter/authorizable_counter.go:
```diff
@@ -0,0 +1,122 @@
+
+// Run callback representing the invocation of a chaincode
+// This chaincode will increment the counter if the user has an attribute called "position" with the value "Software Engineer"
+func (t *AuthorizableCounterChaincode) Run(stub *shim.ChaincodeStub, function string, args []string) ([]byte, error) {
```
The Run function has been deprecated. This should be removed.

andresgaragiola replied:
Removed

srderson commented on examples/chaincode/go/authorizable_counter/authorizable_counter.go:
```diff
@@ -0,0 +1,122 @@
+
+// Query callback representing the query of a chaincode
+func (t *AuthorizableCounterChaincode) Query(stub *shim.ChaincodeStub, function string, args []string) ([]byte, error) {
+	if function != "query" {
```
Naming the function of the Query function "query" is confusing. It would be better to give it a different name.

andresgaragiola replied:
I have called that function "read" now.

srderson commented on membersrvc/ca/aca.go:
```diff
@@ -0,0 +1,543 @@
+		attr.GetID(), attr.GetAffiliation(), attr.GetAttributeKey()).Scan(&count)
+
+	if err != nil {
+		return nil
```
shouldn't this return err?

andresgaragiola replied:
Fixed. Now returns err.

srderson commented on membersrvc/ca/client_grpc.go:
```diff
@@ -0,0 +1,28 @@
+
+)
+
+func GetClientConn(address string, serverName string) (*grpc.ClientConn, error) {
```
Missing comment on public function. golint needs to be run

andresgaragiola replied:
I have added the comment.

srderson commented on membersrvc/ca/client_grpc.go:
```diff
@@ -0,0 +1,28 @@
+	return  grpc.Dial(address, opts...)
+}
+
+func GetACAClient() (*grpc.ClientConn, pb.ACAPClient, error) {
```
Missing comment on public function. golint needs to be run

andresgaragiola replied:
I have added the comment.

srderson commented on membersrvc/ca/eca.go:
```diff
@@ -197,6 +204,10 @@ func (eca *ECA) startECAA(srv *grpc.Server) {
 }
 
+func (eca *ECA) ivokeACAFetchAttributes(id, affiliation string) {
+	//eca.aca.
```
There's no code in this function. What's going on?

andresgaragiola replied:
Removed empty function.

srderson commented on membersrvc/ca/eca.go:
```diff
@@ -314,8 +369,17 @@ func (ecap *ECAP) CreateCertificatePair(ctx context.Context, in *pb.ECertCreateR
 		} else {
 			obcECKey = ecap.eca.obcPub
 		}
+		if role == 1 {
```
See https://github.com/hyperledger/fabric/pull/1444. Using just an int makes the code difficult to read. Use the role type instead.

andresgaragiola replied:
Fixed.

srderson commented on membersrvc/membersrvc.yaml:
```diff
@@ -101,8 +101,18 @@ eca:
-
+                 enabled: true
+aca:
+          attributes:
```
Can you add some comments about what's going on here. I have no clue what represented by the value of these attribute entries.

andresgaragiola replied:
I have added comments

srderson commented on membersrvc/membersrvc.yaml:
```diff
@@ -101,8 +101,18 @@ eca:
+aca:
+          attributes:
+              attribute-entry-0: diego;institution_a;company;IBM;2015-01-01T00:00:00-03:00;;
+              attribute-entry-1: diego;institution_a;position;Software Staff;2015-01-01T00:00:00-03:00;2015-07-12T23:59:59-03:00;
```
This one ends with ';'. The rest end with ';;'. Why?

andresgaragiola replied:
ACA attributes has been added within the membersrvc.yaml temporarily in the future external attribute authority will be invoked.  Each attribute has the format:

  userid;affiliaition;attributename;attributevalue;validfrom;validto;

validTo empty means that the attribute doesn't expire. That is the reason you see ";;" 
Anyway I'm adding a format comment in this file and modifying the company name to a fantasy name.

srderson commented on membersrvc/protos/ca.proto:
```diff
@@ -178,8 +190,9 @@ message ECertCreateReq {
     CertPair certs = 1;
     Token chain = 2;
-    bytes pkchain = 4;
+    bytes pkchain = 5;
```
Why change the number?

andresgaragiola replied:
The number change because a new field was add. Previously pkchain had the value 4 in this way: 

message ECertCreateResp {
    CertPair certs = 1;
    Token chain = 2;
    bytes pkchain = 4;
    Token tok = 3;
   }

I have keep pkchain as the last protobuf field in the same way. I will test it putting pkchain in the third position.

srderson commented on membersrvc/protos/ca.proto:
```diff
@@ -316,3 +329,58 @@ message CertPair {
     bytes enc = 2; // encryption certificate, DER / ASN.1 encoded
 }
+
+message TCertAttributeHash {
```
Comment please

andresgaragiola replied:
I have added the comment.

srderson commented on membersrvc/protos/ca.proto:
```diff
@@ -316,3 +329,58 @@ message CertPair {
+    bytes  attributeValueHash = 2;
+}
+
+message ACAAttrReq {
```
Comment please

andresgaragiola replied:
I have added the comment.

srderson commented on membersrvc/protos/ca.proto:
```diff
@@ -316,3 +329,58 @@ message CertPair {
+    Signature signature = 5; 
+}
+
+message ACAAttrResp {
```
Comment please

andresgaragiola replied:
I have added the comments.

srderson commented on membersrvc/protos/ca.proto:
```diff
@@ -316,3 +329,58 @@ message CertPair {
+    Signature signature = 3; 
+}
+
+message ACAFetchAttrReq {
```
Comment please

andresgaragiola replied:
I have added the comments.

srderson commented on membersrvc/protos/ca.proto:
```diff
@@ -316,3 +329,58 @@ message CertPair {
+    Signature signature = 3; 
+}
+
+message ACAFetchAttrResp {
```
Comment please

andresgaragiola replied:
I have added the comments.

srderson commented on membersrvc/protos/ca.proto:
```diff
@@ -316,3 +329,58 @@ message CertPair {
+    StatusCode status = 1;
+}
+
+message FetchAttrsResult {
```
Comment please

andresgaragiola replied:
I have added the comments.

srderson commented on membersrvc/protos/ca.proto:
```diff
@@ -316,3 +329,58 @@ message CertPair {
+	string Msg = 2;
+}
+
+message ACAAttribute {
```
Comment please

andresgaragiola replied:
I have added the comments.

srderson commented on membersrvc/protos/ca.proto:
```diff
@@ -316,3 +329,58 @@ message CertPair {
+message ACAAttribute {
+    string attributeName = 1;
+    bytes attributeValue = 2;
+	google.protobuf.Timestamp validFrom = 3;
```
Random indent?

srderson commented on protos/chaincode.proto:
```diff
@@ -70,6 +70,7 @@ message ChaincodeSpec {
     string secureContext = 5;
     ConfidentialityLevel confidentialityLevel = 6;
     bytes metadata = 7;
+    map<string, string> attributes = 8;
```
Should the attributes be part of the fabric.Transaction message instead of a ChaincodeSpec?

srderson commented on core/chaincode/shim/crypto/ac/abac.go:
```diff
@@ -0,0 +1,166 @@
+	"crypto/x509"
+	"errors"
+
+	"github.com/hyperledger/fabric/core/crypto/abac"
```
Agree with @muralisrini's comments, the tight coupling will make it very difficult to implement chaincode shims in other languages and is making the binaries too big.

andresgaragiola replied:
Agree with @srderson, @muralisrini and @diegomasini. I understand the problem but in my opinion duplicate code is not a good option. I think that a new issue could be opened to treat this problem in a more general way. I think that a possible solution could be structure a few core libraries to be imported into the shim.

For example attributes have to be processed in three places client, membership services and here. To do that I have defined the abac library, this library is imported from the three places, and defines the core functionality required by abac as for example derivates an attribute key. Probably we could redefine this library to reduce the size and move it to other place, but in my opinion a core library used from the three places is required. Currently is located in the same place that other libraries like "primitives" or "utils" in similar situation.

diegomasini replied:
@srderson @muralisrini as @andresgaragiola mentioned I prefer not to duplicate the code. Could be possible to make it a vendor package? Or that will create other kind of issues?

andresgaragiola committed:
Fixed several findings commented in PR #1513.

andresgaragiola committed:
Merge remote-tracking branch 'upstream/master' into abac

andresgaragiola committed:
Merge remote-tracking branch 'origin/abac' into abac

andresgaragiola committed:
Reverting security enabled, privacy and abac enabled in core.yaml.

mastersingh24 commented:
folks - I have a bunch of suggestions to make here, but they are dependent on agreement that we take this as an opportunity to:

1) Break this PR into manageable chunks  (e.g. membership service piece, chaincode piece, etc)
2) Fix previous "mistakes" not directly part of this PR (e.g. membership services needs to be self testable - tests should not be run out of the core/crypto package)
3) Actually discuss what belongs in the shim and what does not (i.e. I don't think that the shim / stub should provide any of the abac methods directly)
4) Implicit in above, but actually add proper tests in the proper places

IMHO - there is WAY too much of a tight coupling of membership services everywhere.

I am also willing to help with coding tests, etc (so not just complaining here)

If we don't do this now, we are going to be in a heap of trouble down the road

christo4ferris commented:
what @mastersingh24 said. I'm willing to help as well. Letting this in as is will add too much technical debt

andresgaragiola committed:
Added some tests in flag in membersrvc.yaml to enable/disable ACA.

andresgaragiola committed:
Merge remote-tracking branch 'upstream/master' into abac

ghaskins commented:
IIUC, this proposal adds a number of library elements that run in chaincode context.  My preference would be to see chaincode context remain as light as possible and rather, extend the platform ABI to support these types of features and allow the chaincode to call the platform functions.  This would have the following benefits:
1. Other chaincode language environments will not need to reinvent the wheel: we implement it once in the peer and let all chaincode access it via explicit grpcs or metadata on the invoke().
2. It will help to facilitate the feature that I have been asking for for a while now: abstract "identity" away from crypto.  Some identities will be crypto based (tcert signed requests from clients, for example), but some will not / can not (the identity of a chaincode that invokes another).

For 2, I envision perhaps that invoke metadata can carry arbitrary attributes....for tcert based operations, the attributes are derived from the tcert attributes.  For other contexts, they might be derived from elements of the caller.  The main point is that the fabric can be considered an authoritative kernel ala SO_PEERCRED (http://man7.org/linux/man-pages/man7/socket.7.html)

adecaro commented on core/crypto/crypto.go:
```diff
@@ -49,13 +49,13 @@ type Client interface {
 
 	// NewChaincodeExecute is used to execute chaincode's functions.
-	NewChaincodeExecute(chaincodeInvocation *obc.ChaincodeInvocationSpec, uuid string) (*obc.Transaction, error)
+	NewChaincodeExecute(chaincodeInvocation *obc.ChaincodeInvocationSpec, attributes map[string]string, uuid string) (*obc.Transaction, error)
```
Shall we move attributes to *Spec in order to avoid breaking retro-compatibility?

adecaro commented on core/crypto/crypto.go:
```diff
@@ -64,16 +64,13 @@ type Client interface {
 
 	// GetNextTCert gets next available (not yet used) transaction certificate.
-	GetNextTCert() (tCert, error)
+	GetNextTCert(attributes map[string]string) (tCert, error)
```
Shall we avoid to expose internal interfaces like tCert? The corresponding public interface is CertificateHandler. Moreover, what is the difference with GetTCertificateHandlerNext?

adecaro commented on core/crypto/client_impl.go:
```diff
@@ -126,22 +141,22 @@ func (client *clientImpl) GetEnrollmentCertificateHandler() (CertificateHandler,
 	// Return the handler
 	handler := &tCertHandlerImpl{}
-	err = handler.init(client, tCert)
+	err = handler.init(client, tCert.tCert)
```
Shall we have a better naming here?

adecaro commented on core/crypto/client_ks.go:
```diff
@@ -105,10 +105,11 @@ func (ks *keyStore) storeUnusedTCerts(tCerts []tCert) (err error) {
+	for _, tCertBlck := range tCertBlocks {
 		// Insert into UsedTCert
-		if _, err = tx.Exec("INSERT INTO TCerts (cert) VALUES (?)", tCert.GetCertificate().Raw); err != nil {
+		if _, err = tx.Exec("INSERT INTO TCerts (attrhash, cert, prkz) VALUES (?, ?, ?)", tCertBlck.attributesHash, tCertBlck.tCert.GetCertificate().Raw, tCertBlck.tCert.GetPreK0()); err != nil {
```
I guess that PreK0 is a secret key. Shall we store it in an encrypted form?

adecaro commented on core/crypto/crypto_test.go:
```diff
@@ -304,20 +313,144 @@ func TestClientGetAttributesFromTCert(t *testing.T) {
 	}
 
-	attributeBytes, err := deployer.ReadAttribute("company", tcertDER)
+	attributeBytes, err := abac.GetValueForAttribute("company", tcert.GetPreK0(), tcert.GetCertificate())
```
Shall we avoid to expose PreK0 that way? May CertificateHandler have a method GetValueForAttribute(attrName) directly?

andresgaragiola commented:
I'm closing this PR because we are splitting this in others. #1639 is the first of them and is a dependency for the others.
